man/strparse.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id$
.TH STRPARSE 3X "30 September 1992"
.SH NAME
str_parse, str_endparse, str_component, str_separator, str_nextpos
.SH SYNOPSIS
.LP
.nf
.ft B
#include "str.h"
.LP
.ft B
str_h str_parse( str, separ, flags, errnop )
char *str ;
char *separ ;
int flags ;
int *errnop ;
.LP
.ft B
void str_endparse( handle )
str_h handle ;
.LP
.ft B
char *str_component( handle )
str_h handle ;
.LP
.ft B
int str_setstr( handle, newstr )
str_h handle ;
char *newstr ;
.LP
.ft B
int str_separator( handle, separ )
str_h handle ;
char *separ ;
.LP
.ft B
char *str_nextpos( handle )
str_h handle ;
.LP
extern int str_errno ;
.SH DESCRIPTION
.LP
These functions are useful for parsing strings.  In this context
parsing means breaking the string into substrings. The substrings are
separated by a list of possible separator characters.
.LP
.B str_component()
returns successive substrings of the string.
.B str_parse()
creates and initializes a string parser with the string
that will be processed, \fIstr\fR, the list of possible separator
characters, \fIsepar\fR, and flags that control how the parser
works. The \fIflags\fR argument is formed by ORing one or more of
the following constants:
.TP 20
.SB STR_RETURN_ERROR
If something goes wrong return a value that indicates that an error occurred
(e.g. out of memory). The default is for the program to be terminated
with an appropriate error message.
.TP
.SB STR_NULL_START
If \fIstr\fR starts with a separator then a zero-length string will be returned
the first time \fBstr_component()\fR is called.
.TP
.SB STR_NULL_END
If \fIstr\fR ends with a separator then a zero-length string will be returned
by \fBstr_component()\fR when the substrings of \fIstr\fR are exhausted.
.TP
.SB STR_MALLOC
The strings returned by \fBstr_component()\fR will be in malloc'ed memory.
By default the substrings are part of \fIstr\fR.
If this option is not used \fIstr\fR will be modified
by \fBstr_component()\fR.
.LP
Finally, \fBSTR_NOFLAGS\fR may be used to specify no flags.
The \fIerrnop\fR argument points to an integer where the string processing
functions will deposit an error code if an error occurs.
If \fIerrnop\fR
is
.SM NULL
the error codes will be placed in \fIstr_errno\fR.
This is useful only if \fBSTR_RETURN_ERROR\fR is used in \fIflags\fR.
It is possible that \fIstr\fP is
.SM NULL.
In this case, a subsequent
.B str_setstr()
should be used to specify the string to be processed.
.LP
.B str_component()
returns successive substrings from the string associated with the
parser specified by \fIhandle\fR.
.LP
.B str_endparse()
destroys the parser specified by \fIhandle\fR.
.LP
.B str_setstr()
changes the processed string to \fInewstr\fP.
.LP
.B str_separator()
replaces the list of separator characters with \fIsepar\fR.
Processing continues from the current position.
.LP
.B str_nextpos()
returns a pointer to the rest of the string. The previous character
is a separator character (if \fBSTR_MALLOC\fR is not set, then the
previous character is
.SM NUL
).
.SH "RETURN VALUES"
.LP
.B str_parse()
returns a parser handle or
.SM NULL
if something goes wrong and \fIflags\fR & \fBSTR_RETURN_ERROR\fR is true.
Possible \fIstr_errno\fR values:
.RS
.TP 20
.SB STR_ENULLSEPAR
\fIsepar\fR is
.SM NULL
.TP
.SB STR_ENOMEM
the program ran out of memory
.RE
.LP
.B str_component()
returns a pointer to the next substring or
.SM NULL
if something goes wrong and \fIflags\fR & \fBSTR_RETURN_ERROR\fR is true.
.LP
.B str_setstr()
returns
.SB STR_OK
on success or
.SB STR_ERR
on failure.
.LP
.B str_separator()
returns
.SB STR_OK
on success or
.SB STR_ERR
on failure.
.LP
.B str_nextpos()
returns a pointer or
.SM NULL
if the end of string has been reached.
.SH BUGS
.B str_component()
modifies the string unless \fBSTR_MALLOC\fR is
set in the parser.
.LP
There should be only one parser active on a specific string. If there
is more than
one, they all must use the \fBSTR_MALLOC\fR option.



CHANGELOG

This file describes the changes to xinetd. The base version is 2.0.0.

2.0.0: (not released)

2.0.1: (not released)

2.0.2:
	Changes to the Makefile.
	Trivial mods to the usage() function

2.0.3:
	Changes to the Makefile.
	Fixed a bug in the usage() function

2.0.4:
	Distribution versions of the Makefile no longer contain any 
	references to options.opt

2.0.5:
	1) Clarified what an "unlisted RPC service" is in the man page
	2) Fixed a bug in remote_address_check which caused access to be denied
		for all hosts if no_access was set and only_from was not set.
	3) Fixed a bug which caused arbitrary syslog levels if the log_type
		was specified as SYSLOG in the "defaults" entry

2.0.6:
	bug fix in child.c: replaced strx_sprint with strx_print when naming
	interceptor processes


-------------------------------------------------------------------------------

2.1.0: (not released)
	This was mostly a clean-up of 2.0
	List of changes:

	A. The man page has been split into 3 parts:
			xinetd.man			:	man page describing the program
			xinetd.conf.man	:	man page describing the configuration file 
										This file now includes figures about the
										overhead of interception.
			xinetd.log.man		:	man page describing the log file
	B. New service attributes
			rpc_number			: to support unlisted RPC services
			nice					: to set the nice value of forked servers
	C. The IDONLY service flag was added
	D. Now uses the timer library unless NO_TIMERS is defined. If NO_TIMERS
		is not defined, the following are also available:
			a) a new option, -cc, to do periodic consistency checks
			b) timeout for reconfiguration
	E.	Configuration file man page now mentions that access control is 
		based on IP-address instead of domain address.
	F.	The interception code now sets the TCP_NODELAY option.
	G. The timeout when contacting a remote identification server is
		configurable both when the service request has been accepted and 
		when the request is rejected. In the former case it defaults to
		infinity while in the latter it defaults to 30 sec.
	H. The log line ids are now constants in an include file.
	I. The fsma library is no longer used.
	J. Most structure fields were renamed; certain functions were renamed too.
		Access to structure fields is now via macros.
	K. A bug in not restoring the number of descriptors to the soft limit
		was fixed.
	L. The -pid option works
	M. New internal services:
			"servers"		: lists active servers
			"services"		: lists active services

2.1.1:
	xinetd produces more meaningful messages if it can't start logging

2.1.2:
	Fixed 2 bugs:
		1) xinetd would request a port number for unlisted RPC services
			One could get around this by simply specifying the port attribute
			for the service; the specified value would be ignored anyway.
		2) xinetd initialization might fail sometimes in function msg_init()
			syslog logging was specified (i.e. the option -syslog was used)
			The cause of this bug was that the 
				xlog_control( xlog, XLOG_GETFD, &fd ) 
			operation will not fail for xlogs connected to syslog and
			the value of 'fd' would be used in the subsequent fcntl(). That
			value is arbitrary since 'fd' is a local variable.
			If 'fd' did not happen to refer to an open descriptor, the program 
			would terminate since the fcntl() would fail (btw, the success
			of the fcntl() call would be harmless if 'fd' happened to
			refer to an open descriptor).

2.1.3:
	Bug fix:
		only_from/no_access addresses would get inverted on little-endian
		machines when such addresses were specified using the numeric notation
		(for example, 128.138.45.3). This bug was in the numeric_addr function
		which did not convert the result of inet_addr to host-byte-order.
		A work around for the bug would be to use the factorized address 
		notation (for example, 128.138.45.{3})

2.1.4:
	Bug fixes:
		1) in dgram_echo(), sin_len was not being set before the invocation of
			recvfrom
		2) in finger_shutdown(), it was possible for Srdline() to return NULL
			(if the remote end would close the socket without sending anything).
			If the RECORD option was set in the log_on_failure flags, this
			would cause the forked xinetd process which did the recording to
			die since it would try to dereference a NULL pointer.


2.1.5:
	Bug fixes:
		1) in exec_server() service descriptor might be closed when
		   execing server. This made it impossible to start servers
			for 'nowait' services.  The bug occurred only Ultrix version
			4.3a or (probably) later (bug discovered and fix provided
			by doug@seas.smu.edu)
		
		2) for systems that supported supplementary group id's, the
			set_credentials() function did not set those group id's
			(they were being inherited from xinetd). Now initgroups(3)
			is called to set the supplementary group id's properly.
			(bug discovered and fix provided by maf+@osu.edu)

2.1.6:
	Bug fixes:
		1) xinetd will crash after reconfiguration if there is a running
			server for a service that was removed from the configuration and
			which logs on exit.
		2) xinetd forked process falls in infinite loop if identd server
			sends a reply that is missing the ending CR-LF.
			(bug discovered and fix provided by Laurent.Wacrenier@gin.obspm.fr).

	We also change the LOGUSER_SUCCESS_TIMEOUT constant in config.h from
	0 seconds (i.e. infinite timeout) to 30 seconds. This avoids infinite
	waits in case the remote host does not send a RST reply when we attempt
	to connect to the IDENTD port, and there is no server listening at
	that port.


2.1.7:
	Bug fix: the HOST flag in the 'log_on_success' attribute was ignored;
				the code was incorrectly checking if the HOST flag was
				set in the log_on_failure attribute
				(bug discovered by frolich@corrine.cpc.cs.ucf.edu)

2.1.8.1: 
	Added support for TCP redirect to a remote host.
		--bbraun

2.1.8.2: 
	Added support for binding to specific interface.
		--bbraun

2.1.8.3:
	Changed redirect so that a remote port can be specified as well.

libs/src/str/strparse.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

#include "config.h"
#include <string.h>
#include <unistd.h>
#include <stdlib.h>

#include "str.h"
#include "strparse.h"
#include "sio.h"

static int str_errno ;

#define HANDLE_ERROR( flags, retval, errp, errval, msg )    \
	if ( flags & STR_RETURN_ERROR )                 \
	{                                               \
               *errp = errval ;                         \
               return( retval ) ;                       \
        }                                               \
        else                                            \
               terminate( msg );


char *new_string( const char *s )
{
   if ( s )
      return strdup( s ) ;
   else
      return 0;
}

str_h str_parse( register char *str, const char *separ, int flags, int *errnop )
{
	register struct str_handle *hp ;
	int *errp = ( errnop == NULL ) ? &str_errno : errnop ;

	if ( separ == NULL ) {
		HANDLE_ERROR( flags, NULL, errp, STR_ENULLSEPAR,
			"STR str_parse: NULL separator" ) ;
	}

	hp = (struct str_handle *) malloc( sizeof( struct str_handle ) ) ;
	if ( hp == NULL ) {
		HANDLE_ERROR( flags, NULL, errp, STR_ENOMEM,
			"STR str_parse: malloc failed" ) ;
	}

	hp->string = str ;
	hp->pos = str ;
	hp->separator = new_string( separ ) ;
	if ( hp->separator == NULL ) {
		if ( flags & STR_RETURN_ERROR )
		{
			free( (char *) hp ) ;
			*errp = STR_ENOMEM ;
			return( NULL ) ;
		} else {
			terminate( "STR str_parse: malloc failed" ) ;
		}
	}
	
	hp->flags = flags ;
	hp->errnop = errp ;
	hp->no_more = ( str == NULL ) ;
	return( (str_h) hp ) ;
}


void str_endparse( str_h handle )
{
	register struct str_handle *hp = (struct str_handle *) handle ;

	if( hp->separator != NULL )
		free( hp->separator ) ;
	free( (char *) handle ) ;
}


/*
 * Change the string
 */
int str_setstr( str_h handle, char *newstr )
{
        register struct str_handle *hp = (struct str_handle *) handle ;
		
        if ( newstr == NULL ) {
                HANDLE_ERROR( hp->flags, STR_ERR, hp->errnop, STR_ENULLSTRING,
                        "STR str_setstr: NULL string" ) ;
        }
			        
        hp->string = newstr ;
        hp->pos = newstr ;
        hp->no_more = FALSE ;
        return( STR_OK ) ;
}


char *str_component( str_h handle )
{
	register char			*start ;
	register char			*last ;
	register unsigned int		sep_count ;
	char				*retval ;
	int				last_char ;
	register struct str_handle	*hp	= (struct str_handle *) handle ;
	register int			first_call = ( hp->pos == hp->string ) ;

	if ( hp->no_more )
		return( NULL ) ;

	/*
	 * Get number of separator characters.
	 * Find beginning of component.
	 */
	sep_count = strspn( hp->pos, hp->separator ) ;

	/*
	 * If this is the first call, and there are separator characters
	 * at the beginning of the string and the STR_NULL_START flag is set
	 * we return a 0-length string.
	 */
	if ( first_call && ( sep_count > 0 ) && ( hp->flags & STR_NULL_START ))
	{
		start = hp->pos ;
		last = hp->pos ;
	}
	else
	{
		start = hp->pos + sep_count ;

		if ( *start == '\0' )
		{
			last = start ;
			hp->no_more = TRUE ;
			if ( ! ( hp->flags & STR_NULL_END ) )
				return( NULL ) ;
		}
		else
		{
			last = strpbrk( start, hp->separator ) ;
			if ( last == NULL )
				last = start + strlen( start ) ;
		}
	}

	/*
	 * At this point, the following variables must be set:
	 *		start:	beginning of component
	 *		last:   end of component + 1
	 *
	 * If STR_MALLOC is set, allocate space for the new string.
	 *
	 * NOTE: If STR_MALLOC is not set, the processed string is trashed.
	 */
	last_char = *last ;
	if ( hp->flags & STR_MALLOC )
	{
		int len = last - start ;

		retval = malloc( (unsigned)len + 1 ) ;
		if ( retval == NULL ) {
			HANDLE_ERROR( hp->flags, NULL, hp->errnop, STR_ENOMEM,
				"STR str_component: malloc failed" ) ;
		}
		strncpy( retval, start, (size_t)len )[ len ] = '\0' ;
	}
	else
	{
		retval = start ;
		*last = '\0' ;
	}

	/*
	 * Check if last_char is NUL to avoid setting hp->pos past the
	 * end of the string
	 */
	hp->pos = ( last_char == '\0' ) ? last : last+1 ;
	return( retval ) ;
}

